File: app/src/main/java/pe/andy/bookholic/MainActivity.java
package pe.andy.bookholic;

import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.databinding.DataBindingUtil;

import lombok.Getter;
import pe.andy.bookholic.databinding.MainActivityBinding;
import pe.andy.bookholic.model.SearchField;
import pe.andy.bookholic.model.SearchQuery;
import pe.andy.bookholic.model.SortBy;
import pe.andy.bookholic.service.BookSearchService;
import pe.andy.bookholic.ui.BookRecyclerList;
import pe.andy.bookholic.ui.LibraryRecyclerList;
import pe.andy.bookholic.ui.ScrollToTopButton;
import pe.andy.bookholic.ui.SearchDoneSnackBar;

public class MainActivity extends AppCompatActivity {

    @Getter
    MainActivityBinding mBinding;

    public LibraryRecyclerList libraryRecyclerList;
    public BookRecyclerList bookRecyclerList;
    @Getter public ScrollToTopButton scrollToTopButton;
    @Getter public SearchDoneSnackBar searchDoneSnackBar;

    SearchView searchView;

    @Getter public BookSearchService searchService;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_main);

        searchService = new BookSearchService(this);

        libraryRecyclerList = new LibraryRecyclerList(this, this.mBinding, searchService);
        bookRecyclerList = new BookRecyclerList(this, mBinding, searchService);
        scrollToTopButton = new ScrollToTopButton(this.mBinding);
        searchDoneSnackBar = new SearchDoneSnackBar(this.mBinding);

        mBinding.fab.setOnClickListener(view -> {
            if (searchView.hasFocus()) {
                searchView.setQuery(searchView.getQuery(), true);
            }
            else {
                mBinding.nestedScrollView.scrollTo(0, 0);
                searchView.setIconified(false);
            }
        });

        // 테스트를 위한 리스트
        //bookRecyclerUi.add(generateTestBooks());
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.search_menu, menu);

        MenuItem searchItem = menu.findItem(R.id.action_search);
        searchView = (SearchView) searchItem.getActionView();
        searchView.setQueryHint("검색어");
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                Log.d("BookHolic", "onQueryTextSubmit");
                searchView.clearFocus();

                SearchQuery sQuery = new SearchQuery(
                        query.trim(),
                        SearchField.TITLE,
                        SortBy.TITLE);

                searchService.search(sQuery);
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                return false;
            }
        });
        return true;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        searchService.cancelAll();
    }

}

File: app/src/main/java/pe/andy/bookholic/MainActivity.kt
package pe.andy.bookholic

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import by.kirich1409.viewbindingdelegate.viewBinding
import pe.andy.bookholic.databinding.ActivityMainBinding
import pe.andy.bookholic.fragment.FavoriteFragment
import pe.andy.bookholic.fragment.SearchFragment

class MainActivity : AppCompatActivity(R.layout.activity_main) {
    val binding: ActivityMainBinding by viewBinding(ActivityMainBinding::bind)

    val searchFragment: SearchFragment = SearchFragment()
    val favoriteFragment: FavoriteFragment = FavoriteFragment()
    var activeFragment: Fragment = searchFragment

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (savedInstanceState == null) {
            setupBottomNavigationBar()
        }
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        setupBottomNavigationBar()
    }

    private fun setupBottomNavigationBar() {
        supportFragmentManager.beginTransaction()
            .add(R.id.host_fragment, favoriteFragment, "favoriteFragment")
            .hide(favoriteFragment)
            .commit()

        supportFragmentManager.beginTransaction()
            .add(R.id.host_fragment, searchFragment, "searchFragment")
            .commit()

        binding.bottomNavigation.setOnItemSelectedListener {
            when (it.itemId) {
                R.id.bottom_navigation_search -> {
                    if (activeFragment != searchFragment) {
                        supportFragmentManager.beginTransaction()
                            .show(searchFragment)
                            .hide(favoriteFragment)
                            .commit()

                        activeFragment = searchFragment
                    }
                }

                R.id.bottom_navigation_favorite -> {
                    if (activeFragment != favoriteFragment) {
                        supportFragmentManager.beginTransaction()
                            .show(favoriteFragment)
                            .hide(searchFragment)
                            .commit()

                        activeFragment = favoriteFragment
                    }
                }
            }
            true
        }
    }

    override fun onBackPressed() {
        when (supportFragmentManager.fragments.lastOrNull()) {
            searchFragment -> {
                super.onBackPressed()
            }
            favoriteFragment -> {
                supportFragmentManager.beginTransaction()
                    .show(favoriteFragment)
                    .hide(searchFragment)
                    .commit()

                activeFragment = searchFragment

                binding.bottomNavigation.menu.getItem(0).isChecked = true
            }
        }
    }
}